include/volk/volk_alloc.h

```diff
@@ -0,0 +1,80 @@
+/* -*- C++ -*- */
+/*
+ * Copyright 2019 Free Software Foundation, Inc.
+ *
+ * This file is part of GNU Radio
+ *
+ * GNU Radio is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3, or (at your option)
+ * any later version.
+ *
+ * GNU Radio is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GNU Radio; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#ifndef INCLUDED_VOLK_ALLOC_H
+#define INCLUDED_VOLK_ALLOC_H
+
+#include <cstdlib>
+#include <limits>
+#include <new>
+#include <vector>
+
+#include <volk/volk.h>
+
+namespace volk {
+
+/*!
+ * \brief C++11 allocator using volk_malloc and volk_free
+ *
+ * \details
+ *   adapted from https://en.cppreference.com/w/cpp/named_req/Alloc
+ */
+template <class T>
+struct alloc {
+  typedef T value_type;
+
+  alloc() = default;
+
+  template <class U> constexpr alloc(alloc<U> const&) noexcept {}
+
+  T* allocate(std::size_t n) {
+    if (n > std::numeric_limits<std::size_t>::max() / sizeof(T)) throw std::bad_alloc();
+
+    if (auto p = static_cast<T*>(volk_malloc(n*sizeof(T), volk_get_alignment())))
+      return p;
+
+    throw std::bad_alloc();
+  }
+
+  void deallocate(T* p, std::size_t) noexcept { volk_free(p); }
+
+} ;
+
+template <class T, class U>
```

Comment:
What do these operators actually do?

Reply:
This code is taken from https://en.cppreference.com/w/cpp/named_req/Allocator and adapted to use `volk_alloc`/`volk_free` instead of `malloc`, `free`. As for `operator==` and `operator!=`, see the explanation in https://stackoverflow.com/questions/47186641/why-c-custom-allocator-needs-comparison-operators

Reply:
About `operator==` ... wow. That's clear as mud. I guess it's required, and that's the syntax ... so that's just the way it is! Do you have a better explanation? Just curious now ...

Reply:
From  https://en.cppreference.com/w/cpp/named_req/Allocator:
![Screenshot from 2019-10-28 09-19-01](https://user-images.githubusercontent.com/7550829/67662620-13752180-f964-11e9-941f-26f677a4d098.png)

So the way `operator==` is defined is saying that any object allocated with `volk_malloc` can be deallocated with `volk_free`.

Reply:
Ah OK now -that- makes sense! And so `operator!=` means if `volk_malloc` didn't do the allocation then `volk_free` can't deallocate it? Or something like that. Makes sense too.

Reply:
> Ah OK now -that- makes sense! And so `operator!=` means if `volk_malloc` didn't do the allocation then `volk_free` can't deallocate it? Or something like that. Makes sense too.

No, `operator!=` is just the logical negation of `operator==`.

Here (https://en.cppreference.com/w/cpp/memory/allocator/operator_cmp) a condition is given for when `operator==`, `operator!=` always return `true` and `false`: this is the case for **stateless** allocators, which I think is true for `volk_malloc`, `volk_free`, and for  `volk_get_alignment`.